Doc: Add docs for running PyInstaller with Python optimizations

nejch opened the pull request:
Closes #2447.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
 
 
+Running PyInstaller with Python optimizations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
The underlining must be at least as long (in number of characters) as the heading.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+.. Note::
+
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
+	this option. Using ``-O``, ``__debug__`` is set to ``False`` and
```
"When using …"

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
+	this option. Using ``-O``, ``__debug__`` is set to ``False`` and
+	``assert`` statements are not executed. The ``-OO`` option
```
assert statements are not only "not executed", but completely removed from the byte-code.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+
+.. Note::
+
+	You should be aware of the ``PYTHONOPTIMIZE`` mechanism when using
```
* Well, this is not an "option" for pyinstaller, so I suggest you using different wording.
* I'd start the sentence with the "When using this …"
* "be aware of" sound like: you should know there is such a think. But one needs no know what this does.
* ``PYTHONOPTIMIZE`` is the name of the env-var. Maybe better just "Python Bytecode Optimization mechanism"

Suggestion:
> When using this feature/option/thing, you should be aware of how the the ``PYTHONOPTIMIZE`` mechanism works.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+
+    python -O -m PyInstaller myscript.py
+
+You can use any PyInstaller options that are otherwise available with the ``pyinstaller`` command. For example::
```
long line.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+PyInstaller can be run with Python optimization options (``-O`` or ``-OO``)
+by executing it as a Python module, rather than using the ``pyinstaller`` command::
+
+    python -O -m PyInstaller myscript.py
```
Want to add these, too?
```
python -OO -m PyInstaller myscript.py
PYTHONOPTIMIZE=1 pyinstaller myscript.py
```

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+
+Alternatively, you can also use the path to pyinstaller::
+
+    python -O /path/to/pyinstaller/ myscript.py
```
I assume there should be no trailing slash. Maybe use `/path/to/bin/pyinstaller`?

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
         myscript.spec
 
 
+Running PyInstaller with Python optimizations
```
PyInstaller should be written `|PyInstaller|` anywhere.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,30 @@ Or in Windows, use the little-known BAT file line continuation::
+	this option. Using ``-O``, ``__debug__`` is set to ``False`` and
+	``assert`` statements are not executed. The ``-OO`` option
+	additionally omits docstrings. If your code relies on any of these
+	features, your program may break or have unexpected behavior.
```
We should add something like "Using this feature/option/thing effects not only your main script but *all* modules compiled by |PyInstaller|"

nejch commented:
Thanks for the feedback, sorry I wasn't aware of the `[skip ci]` tag. I've added the changes requested and built it with sphinx to test. I changed the option wording to 'feature' (and 'flag' for `-O`) to avoid confusion with PyInstaller options.

The snippets might a bit wordy now as I added a separate Windows line for running with `PYTHONOPTIMIZE=1`, see if it's ok though. I'll push the changes in a minute.

htgoebel commented on doc/usage.rst:
```diff
@@ -89,6 +89,49 @@ Or in Windows, use the little-known BAT file line continuation::
+    The ``-OO`` flag additionally removes docstrings.
+
+    Using this feature affects not only your main script, but *all* modules
+    compiled by |PyInstaller|. If your code (or any module imported by your
```
This should be "included by PyInstaller", since PyInstaller will not compile the source if there already is a .pyo file.

htgoebel commented:
Thanks for the update. Sorry for not applying it earlier.
